felix/calc/l3_route_resolver.go

```diff
@@ -332,27 +337,47 @@ func (c *L3RouteResolver) OnResourceUpdate(update api.Update) (_ bool) {
 	var nodeInfo *l3rrNodeInfo
 	if update.Value != nil {
 		node := update.Value.(*apiv3.Node)
-		if node.Spec.BGP != nil && node.Spec.BGP.IPv4Address != "" {
+		if node.Spec.BGP != nil && (node.Spec.BGP.IPv4Address != "" || node.Spec.BGP.IPv6Address != "") {
 			bgp := node.Spec.BGP
-			// Use cnet.ParseCIDROrIP so we get the IP and the CIDR.  The parse functions in the ip package
-			// throw away one or the other.
-			ipv4, caliNodeCIDR, err := cnet.ParseCIDROrIP(bgp.IPv4Address)
-			if err != nil {
-				logrus.WithError(err).Panic("Failed to parse already-validated IP address")
+			nodeInfo = &l3rrNodeInfo{}
+			if bgp.IPv4Address != "" {
+				// Use cnet.ParseCIDROrIP so we get the IP and the CIDR.  The parse functions in the ip package
+				// throw away one or the other.
+				ipv4, caliNodeCIDR, err := cnet.ParseCIDROrIP(bgp.IPv4Address)
+				if err != nil {
+					logrus.WithError(err).Panic("Failed to parse already-validated IPv4 address")
+				}
+				nodeInfo.V4Addr = ip.FromCalicoIP(*ipv4).(ip.V4Addr)
+				nodeInfo.V4CIDR = ip.CIDRFromCalicoNet(*caliNodeCIDR).(ip.V4CIDR)
 			}
-			nodeInfo = &l3rrNodeInfo{
-				Addr: ip.FromCalicoIP(*ipv4).(ip.V4Addr),
-				CIDR: ip.CIDRFromCalicoNet(*caliNodeCIDR).(ip.V4CIDR),
+			if bgp.IPv6Address != "" {
+				ipv6, caliNodeCIDRV6, err := cnet.ParseCIDROrIP(bgp.IPv6Address)
+				if err != nil {
+					logrus.WithError(err).Panic("Failed to parse already-validated IPv6 address")
+				}
+				nodeInfo.V6Addr = ip.FromCalicoIP(*ipv6).(ip.V6Addr)
+				nodeInfo.V6CIDR = ip.CIDRFromCalicoNet(*caliNodeCIDRV6).(ip.V6CIDR)
 			}
 		} else {
-			ipv4, caliNodeCIDR := cresources.FindNodeAddress(node, apiv3.InternalIP)
+			ipv4, caliNodeCIDR := cresources.FindNodeAddress(node, apiv3.InternalIP, 4)
+			ipv6, caliNodeCIDRV6 := cresources.FindNodeAddress(node, apiv3.InternalIP, 6)
```

Comment:
nit, but I'd move this call for v6 down to line 367 - best to keep variable declarations as close to the first place they are used as possible.

Reply:
Will do!